src/esp/gfx/BatchRenderer.cpp

```diff
@@ -0,0 +1,594 @@
+// Copyright (c) Facebook, Inc. and its affiliates.
+// This source code is licensed under the MIT license found in the
+// LICENSE file in the root directory of this source tree.
+
+#include "BatchRenderer.h"
```

Comment:
1. No, the use case is too narrow and restricted to be useful for any other Magnum users except Habitat itself. Plus, as it's quite high-level compared to for example the multi-draw-enabled shaders, having it here means faster iteration times and easier experiments and perf tuning than the complex cycle of implementing, testing, documenting and upstreaming a feature in Magnum and then using it here. That said, what's here will tend to get smaller over time, as I integrate the lower-level bits of it into Magnum -- such as hierarchic transformation calculation, builtin mesh view support and such.
2. I wouldn't discourage that, of course in moderation, not like pulling the whole Habitat JSON config parsing in here :) But -- as we discussed several times before -- I can imagine the Habitat "client" and the renderer sharing certain data such as bounding volume annotations for both culling and physics or debug visualizers, and then this would pull in common helpers to retrieve such data consistently on both sides. Such data could also be duplicated in Habitat-specific structures, but to me that feels like wasteful extra work if it would only be to have the renderer not rely on any Habitat code.

Thinking further, I intend to make this batch renderer replace / take over everything that's in the `gfx` library, and then it'll be as you say -- a standalone "just Magnum code" library that shares very little with the rest of Habitat.

Reply:
What I'm specifically trying to avoid/discourage: some hacky Habitat-specific renderer "features" like toggling lights, debug-drawing, etc. We want this to be reusable with other simulators in the future. I guess we can leave this here in `gfx` and then just try to be vigilant.

Reply:
I think both toggling lights and debug drawing is a general-enough feature that would cause no problem here. *If* it's made in a way that doesn't negatively affect perf in scenarios that don't need them, of course. 

What I would see highly problematic is dragging Habitat config JSON parsing into here, for example. But as long as I'm in charge of this code, that won't happen ;)

Reply:
Ok, poking this a bit more with a concrete example: Imagine we call `LightSetup.h getDefaultLights` in this file. `LightSetup.h` includes `SceneGraph.h`, which includes `esp/sensor/VisualSensor.h` and `SceneNode.h`. At this point, all this Habitat code is now hopelessly intertwined and it would become hard to use this batch renderer with another simulator (I'm thinking upcoming batch Mujoco specifically).

So how should all this actually work? LightSetup or other Habitat code should call into this file, e.g. calling `BatchRenderer::addLight`.

This is my motivation to separate this from `gfx` lib: to cut off access to those headers and thus force this dependency relationship.

Reply:
Ah, now I see what you mean ... especially given the current state of "everything includes everything else" :sweat_smile: 

Yep, making it a separate library would cut off the access, at least until someone needs to do a permanent temporary hack that links everything else to it again. But then a cunterexample -- wouldn't it make sense to eventually use Habitat's own logging routines in here? For Python-aware assertions, filtering and all that. Which is in the `core` library, which however unfortunately somehow pulls in the JSON headers as well, as far as I can see, and we're back at the start :/ I'm afraid the only viable long-term solution is to start enforcing header/dependency hygiene across the whole project...

But you have a point, and this is an opportunity to make the dependencies clean from the start. Unlike the rest of `gfx` I don't need any physics or Eigen integration here, so should I put this into a new `batched_gfx` library instead? Ideas for a fancier / shorter name?

Reply:
>But then a cunterexample -- wouldn't it make sense to eventually use Habitat's own logging routines in here?
Agree this is tough. When we get to that point (adding logging), we can either (1) add the dependency on `core`, (2) do non-Habitat logging in here, or (3) clean up and separate `core` into more reasonable pieces.